File: state_space/deepc.py
import numpy as np
import cvxpy as cp

class DeePC_Controller:
    def __init__(self, model, delay):
        self.T_prev = 6
        self.T_fut = 15
        self.L = self.T_prev + self.T_fut
        self.T = 100
        
        self.Q = np.eye(model.ny * self.T_fut)
        self.R = np.eye(model.nu * self.T_fut) * 0.1
        self.slack_cost = 1000
        
        u_sim = 2 * np.random.randn(model.nu, self.T)
        x_sim = np.zeros((model.nx, self.T+1))
        y_sim = np.zeros((model.ny, self.T))
        
        for i in range(self.T):
            u_index = i - delay
            u = u_sim[:, u_index] if u_index >= 0 else np.zeros(model.nu)
            x_sim[:, i+1] = model.x_next(x_sim[:, i], u)
            y_sim[:, i] = model.y(x_sim[:, i])
        
        self.num_hankel_columns = self.T - self.L + 1
        
        self.Hu = np.zeros((model.nu * self.L, self.num_hankel_columns))
        self.Hy = np.zeros((model.ny * self.L, self.num_hankel_columns))
        
        for i in range(self.num_hankel_columns):
            self.Hu[:, i] = u_sim[:, i:i+self.L].T.reshape(-1)
            self.Hy[:, i] = y_sim[:, i:i+self.L].T.reshape(-1)          
    
        
    def step(self, model, x0, x_ss, u_ss, u_past_, y_past_, delay = 0, prev_inputs = [0], current_index = 0):
        U_p = self.Hu[:model.nu*self.T_prev]
        U_f = self.Hu[-model.nu*self.T_fut:]
        Y_p = self.Hy[:model.ny*self.T_prev]
        Y_f = self.Hy[-model.ny*self.T_fut:]
        
        u_past = np.zeros((model.nu, self.T_prev))
        y_past = np.zeros((model.ny, self.T_prev))
        
        for i in range(self.T_prev):
            past_index = current_index - self.T_prev + i
            if past_index >= 0:
                u_past[:, i] = u_past_[:, past_index]
                y_past[:, i] = y_past_[:, past_index]
                
        u_past = u_past.T.reshape(-1)
        y_past = y_past.T.reshape(-1)
                
        y_ref = model.y(x_ss)
        y_ref = np.tile(y_ref, self.T_fut)
        u_ref = np.tile(u_ss, self.T_fut)
        
        g = cp.Variable(self.num_hankel_columns)
        u = cp.Variable(model.nu*self.T_fut)
        y = cp.Variable(model.ny*self.T_fut)
        s = cp.Variable(model.nu*self.T_fut)
        
        cost = cp.quad_form(y - y_ref, self.Q) + cp.quad_form(u - u_ref, self.R) + cp.norm1(s)
        
        constraints = [
            U_p @ g == u_past,
            U_f @ g == u,
            Y_p @ g == y_past,
            Y_f @ g == y,
            cp.norm_inf(u) <= 3,
            s >= 0
        ]
        
        prob = cp.Problem(cp.Minimize(cost), constraints)
        
        result = prob.solve()
        
        if np.isinf(result):
            return np.zeros(model.nu)
        
        return u.value[:model.nu]
    
if __name__ == '__main__':
    from model import LinearStateSpaceModel
    
    model = LinearStateSpaceModel(np.eye(2), np.eye(2))
    DeePC_Controller(model)

File: src/controllers/deepc.cpp
#include "deepc.hpp"

DeePC::DeePC(const std::shared_ptr<LTIStateSpaceModel> &model) : model_(std::move(model))
{
    ny_ = model_->ny;
    nu_ = model_->nu;

    T_prev_ = 6;
    T_fut_ = 15;
    L_ = T_prev_ + T_fut_;
    int T_traj = 100;

    Q_ = Matrix::Identity(ny_, ny_);
    R_ = Matrix::Identity(nu_, nu_) * 0.1;

    Matrix u_sim = 2 * Matrix::Random(nu_, T_traj);
    Matrix x_sim = Matrix::Zero(model_->nx, T_traj + 1);
    Matrix y_sim = Matrix::Zero(ny_, T_traj);

    for (int i = 0; i < T_traj; i++)
    {
        Vector u = u_sim.col(i);
        x_sim.col(i + 1) = model_->x_next(x_sim.col(i), u);
        y_sim.col(i) = model->y<Vector>(x_sim.col(i));
    }

    num_hankel_columns_ = T_traj - L_ + 1;

    Hu_ = Matrix::Zero(nu_ * L_, num_hankel_columns_);
    Hy_ = Matrix::Zero(ny_ * L_, num_hankel_columns_);

    for (int i = 0; i < num_hankel_columns_; i++)
    {
        Hu_.col(i) = u_sim.block(0, i, nu_, L_).reshaped(nu_ * L_, 1);
        Hy_.col(i) = y_sim.block(0, i, ny_, L_).reshaped(ny_ * L_, 1);
    }

    num_vars_ = num_hankel_columns_ + (ny_ + nu_) * T_fut_;
    num_constraints_ = (ny_ + nu_) * L_ + nu_ * T_fut_;

    solver_.settings()->setVerbosity(false);
    solver_.data()->setNumberOfVariables(num_vars_);
    solver_.data()->setNumberOfConstraints(num_constraints_);
}

Vector DeePC::step(Vector x0, Vector x_ss, Vector u_ss, Matrix u_past, Matrix y_past, int current_index)
{
    solver_.clearSolver();
    solver_.data()->clearHessianMatrix();
    solver_.data()->clearLinearConstraintsMatrix();

    /* Cost (3g 2y 2u):
    H: 0 0 0 0 0 0 0 -- g no cost
       0 0 0 0 0 0 0
       0 0 0 0 0 0 0
       0 0 0 Q 0 0 0 -- y costs
       0 0 0 0 Q 0 0
       0 0 0 0 0 R 0 -- u costs
       0 0 0 0 0 0 R
    */

    Matrix H = Matrix::Zero(num_vars_, num_vars_);
    Vector g = Vector::Zero(num_vars_);
    Vector y_ref = model_->y(x_ss);

    for (int i = 0; i < T_fut_; i++)
    {
        H.block(num_hankel_columns_ + i * ny_, num_hankel_columns_ + i * ny_, ny_, ny_) = Q_;
        H.block(num_hankel_columns_ + T_fut_ * ny_ + i * nu_, num_hankel_columns_ + T_fut_ * ny_ + i * nu_, nu_, nu_) = R_;

        g.segment(num_hankel_columns_ + i * ny_, ny_) = -Q_ * y_ref;
        g.segment(num_hankel_columns_ + T_fut_ * ny_ + i * nu_, nu_) = -R_ * u_ss;
    }

    /* Constraints:
    A: Y_p 0  0
       Y_f -I 0
       U_p 0  0
       U_f 0  -I
    b: y_past
       0
       u_past
       0
    */

    Matrix Y_p = Hy_.topRows(ny_ * T_prev_);
    Matrix Y_f = Hy_.bottomRows(ny_ * T_fut_);
    Matrix U_p = Hu_.topRows(nu_ * T_prev_);
    Matrix U_f = Hu_.bottomRows(nu_ * T_fut_);

    Matrix u_past_block = Matrix::Zero(nu_, T_prev_);
    Matrix y_past_block = Matrix::Zero(ny_, T_prev_);

    for (int i = 0; i < T_prev_; i++)
    {
        int past_index = current_index - T_prev_ + i;
        if (past_index >= 0)
        {
            u_past_block.col(i) = u_past.col(past_index);
            y_past_block.col(i) = y_past.col(past_index);
        }
    }

    Vector u_past_vec = u_past_block.reshaped(nu_ * T_prev_, 1);
    Vector y_past_vec = y_past_block.reshaped(ny_ * T_prev_, 1);

    Matrix dynamics_constr_A = Matrix::Zero(num_constraints_, num_vars_);
    Vector dynamics_constr_u = Vector::Zero(num_constraints_);
    Vector dynamics_constr_l = Vector::Zero(num_constraints_);

    // y data constraints
    dynamics_constr_A.block(0, 0, ny_ * T_prev_, num_hankel_columns_) = Y_p;
    dynamics_constr_A.block(ny_ * T_prev_, 0, ny_ * T_fut_, num_hankel_columns_) = Y_f;
    dynamics_constr_A.block(ny_ * T_prev_, num_hankel_columns_, ny_ * T_fut_, ny_ * T_fut_) = -Matrix::Identity(ny_ * T_fut_, ny_ * T_fut_);

    dynamics_constr_l.head(ny_ * T_prev_) = y_past_vec;
    dynamics_constr_u.head(ny_ * T_prev_) = y_past_vec;

    // u data constraints
    dynamics_constr_A.block(ny_ * L_, 0, nu_ * T_prev_, num_hankel_columns_) = U_p;
    dynamics_constr_A.block(ny_ * L_ + nu_ * T_prev_, 0, nu_ * T_fut_, num_hankel_columns_) = U_f;
    dynamics_constr_A.block(ny_ * L_ + nu_ * T_prev_, num_hankel_columns_ + ny_ * T_fut_, nu_ * T_fut_, nu_ * T_fut_) = -Matrix::Identity(nu_ * T_fut_, nu_ * T_fut_);

    dynamics_constr_l.segment(ny_ * L_, nu_ * T_prev_) = u_past_vec;
    dynamics_constr_u.segment(ny_ * L_, nu_ * T_prev_) = u_past_vec;

    // Additional constraints
    dynamics_constr_A.block((ny_ + nu_) * L_, num_hankel_columns_ + ny_ * T_fut_, nu_ * T_fut_, nu_ * T_fut_) = Matrix::Identity(nu_ * T_fut_, nu_ * T_fut_);
    dynamics_constr_l.segment((ny_ + nu_) * L_, nu_ * T_fut_) = Vector::Constant(nu_ * T_fut_, -3);
    dynamics_constr_u.segment((ny_ + nu_) * L_, nu_ * T_fut_) = Vector::Constant(nu_ * T_fut_, 3);

    // Solver Configuration
    solver_.data()->setHessianMatrix((Eigen::SparseMatrix<double>)H.sparseView());
    solver_.data()->setGradient(g);
    solver_.data()->setLinearConstraintsMatrix((Eigen::SparseMatrix<double>)dynamics_constr_A.sparseView());
    solver_.data()->setBounds(dynamics_constr_l, dynamics_constr_u);

    solver_.initSolver();
    solver_.solveProblem();

    if (solver_.getStatus() == OsqpEigen::Status::Solved)
    {
        Vector sol = solver_.getSolution();
        return sol.segment(num_hankel_columns_ + T_fut_ * ny_, nu_);
    }
    return Vector::Zero(nu_);
}